util/udma_barrier.h

```diff
@@ -194,10 +194,29 @@
 #define mmio_flush_writes() asm volatile("membar #StoreStore" ::: "memory")
 #elif defined(__aarch64__)
 #define mmio_flush_writes() asm volatile("dsb st" ::: "memory");
-#elif defined(__sparc__) || defined(__s390x__)
+#elif defined(__sparc__)
 #define mmio_flush_writes() asm volatile("" ::: "memory")
 #elif defined(__loongarch__)
 #define mmio_flush_writes() asm volatile("dbar 0" ::: "memory")
+#elif defined(__s390x__)
+#include <sys/auxv.h>
+
+static inline void s390_pciwb(void)
+{
+#ifdef HWCAP_S390_PCI_MIO
+	/* As udma_barrier.h is stand alone we can't access
+	   the s390_mio_enabled variable used in mmio.h. Instead just check
+	   the ELF HWCAP directly which in testing seems to have very low
+	   overhead.
+	 */
+	if (!(getauxval(AT_HWCAP) & HWCAP_S390_PCI_MIO))
```

Comment:
This should be preloaded too

Reply:
This is also only meaningful with WC memory, does s390 have that?

Reply:
Yes we do. Do you have a suggestion for where to put the preloading?

Reply:
You might just do all of this like we did sse on x86 with HAVE_FUNC_ATTRIBUTE_IFUNC

Reply:
I'll look into it. I'd assume since this is a linker thing it can't be used in a header inline function though because they get inline before the code is been by the linker so I would still need a `.c` file, no?

Reply:
Yes, like we did for sse

Reply:
You'd do it for s390_mmio_Write and inline the pcistgi stuff in the C file

Reply:
Ok sure that's doable but what would I do here for the `mmio_flush_write()`?

Reply:
Same thing, just one implementation is a NOP

Reply:
I meant more which C file to put it in. The `udma_barrier.h` is also directly included in a compile test snippet for `HAVE_COHERENT_DMA` so that would then also have to compile that in.

Reply:
Just make a new one for this s390 stuff. The compile test may need to be changed to avoid linking.

Reply:
@jgunthorpe I'm still wondering about IFUNC attribute vs a flag like in this version. With the flag I understand the disadvantage is the global variable I would however guess that it has less overhead because all the `readNN_be/le()` functions are inlined with a known length at compile-time. Also we don't have the function call overhead with the flag boiling down to a single load that as you correctly pointed out is likely well predicted by the branch predictor. How would you feel about the following setup.
* S390 PCI instruction wrappers in `util/s390_mmio.h` and the `MAKE_READ/WRITE` macros. This would be included in `util/mmio.h` and `util/udma_barrier.h` and would use a flag. As you said we need to avoid linking in the compile test either way.
* S390 `mmio_memcpy_x64()` in `util/mmio.c` either with IFUNC or the flag, here the function call overhead is spread over a larger data size and we usually don't know the length at compile time so inlining is less important. Personally I find the IFUNC harder to read than a flag, at least when it's not available as an attribute. Either way I don't think we need to have `mmio_memcpy_x64()` in a header.